examples/http-proxy/proxy.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		die(err)
+	}
+	listen, err := ma.NewMultiaddr(fmt.Sprintf("/ip4/127.0.0.1/tcp/%d", port))
+	if err != nil {
+		die(err)
+	}
+	ps := ps.NewPeerstore()
+	ps.AddPrivKey(pid, priv)
+	ps.AddPubKey(pid, pub)
+	n, err := swarm.NewNetwork(context.Background(),
+		[]ma.Multiaddr{listen}, pid, ps, nil)
+	if err != nil {
+		die(err)
+	}
+	return bhost.New(n)
+}
+
+func log(err error) {
+	fmt.Fprintf(os.Stderr, "%s\n", err)
+}
+
+func die(err error) {
+	log(err)
+	os.Exit(1)
+}
+
+// ProxyService provides HTTP proxying on top of libp2p by launching an
+// HTTP server which tunnels the requests to a destination peer running
+// ProxyService too.
+type ProxyService struct {
+	host      host.Host
+	dest      peer.ID
+	proxyAddr ma.Multiaddr
+}
+
+// NewProxyService attaches a proxy service to the given libp2p Host.
+// The proxyAddr parameter specifies the address on which the
+// HTTP proxy server listens. The dest parameter specifies the peer
+// ID of the remote peer in charge of performing the HTTP requests.
+//
+// ProxyAddr/dest may be nil/"" it is not necessary that this host
+// provides a listening HTTP server (and instead its only function is to
+// perform the proxied http requests it receives from a different peer.
+//
+// The addresses for the dest peer should be part of the host's peerstore.
+func NewProxyService(h host.Host, proxyAddr ma.Multiaddr, dest peer.ID) *ProxyService {
+	// We let our host know that it needs to handle streams tagged with the
+	// protocol id that we have defined, and then handle them to
+	// our own streamHandling function.
+	h.SetStreamHandler(Protocol, streamHandler)
+
+	fmt.Println("Proxy server is ready")
+	fmt.Println("libp2p-peer addresses:")
+	for _, a := range h.Addrs() {
+		fmt.Printf("%s/ipfs/%s\n", a, peer.IDB58Encode(h.ID()))
+	}
+
+	return &ProxyService{
+		host:      h,
+		dest:      dest,
+		proxyAddr: proxyAddr,
+	}
+}
+
+// streamHandler is our function to handle any libp2p-net streams that belong
+// to our protocol. The streams should contain an HTTP request which we need
+// to parse, make on behalf of the original node, and then write the response
+// on the stream, before closing it.
+func streamHandler(stream inet.Stream) {
+	// Remember to close the stream when we are done.
+	defer stream.Close()
+
+	// Create a new buffered reader, as ReadRequest needs one.
+	// The buffered reader reads from our stream, on which we
+	// have sent the HTTP request (see ServeHTTP())
+	buf := bufio.NewReader(stream)
+	// Read the HTTP request from the buffer
+	req, err := http.ReadRequest(buf)
+	if err != nil {
+		log(err)
+		return
+	}
+	defer req.Body.Close()
+
+	// We need to reset these fields in the request
+	// URL as they are not maintained.
+	req.URL.Scheme = "http"
+	hp := strings.Split(req.Host, ":")
+	if len(hp) > 1 && hp[1] == "443" {
+		req.URL.Scheme = "https"
+	} else {
+		req.URL.Scheme = "http"
+	}
+	req.URL.Host = req.Host
+
+	outreq := new(http.Request)
+	*outreq = *req
```

Comment:
nah, need to pass *http.Request to RoundTrip. I'll leave it as it is, in case someone wants to improve it borrowing more stuff from the `reverseproxy.go`.